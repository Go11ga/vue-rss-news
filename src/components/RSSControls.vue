<template>
    <div class="rss-controls">
        <div class="rss-controls__item">

            <router-link
                class="rss-controls__link" 
                to="/all/1"
                :class="{ 'rss-controls__link--active' : getActiveChannel === 'all' }"
            >
                All
            </router-link>

            <router-link
                class="rss-controls__link" 
                to="/mos/1"
                :class="{ 'rss-controls__link--active' : getActiveChannel === 'mos' }"
            >
                Mos
            </router-link>
            <router-link
                class="rss-controls__link" 
                to="/ria/1"
                :class="{ 'rss-controls__link--active' : getActiveChannel === 'ria' }"
            >
                RIA
            </router-link>

        </div>

        <div class="rss-controls__item">
            <button 
                class="rss-controls__btn"
                :class="{'rss-controls__btn--active' : view === 'list'}"
                @click="toggleView('list')"
            >
                <svg class="rss-controls__icon">
                    <use xlink:href="#icon_list"></use>
                </svg>
            </button>

            <button 
                class="rss-controls__btn"
                :class="{'rss-controls__btn--active' : view === 'grid'}"
                @click="toggleView('grid')"
            >
                <svg class="rss-controls__icon">
                    <use xlink:href="#icon_grid"></use>
                </svg>
            </button>
        </div>
        
    </div>
</template>


<script>
import { mapGetters, mapMutations } from 'vuex'

export default {
    name: 'RSSControls',

    data: () => ({
        view: 'grid'
    }),

    computed: {
        ...mapGetters([
            'getActiveChannel'
        ])
    },

    methods: {
        ...mapMutations([
            'setViewMode'
        ]),

        toggleView(val) {
            if (this.view === val) {
                return 
            }

            this.view = val

            localStorage.setItem('RSSView', val)

            this.setViewMode(val)
        }
    },

    created() {
        let viewMode = localStorage.getItem('RSSView') || 'grid'

        this.view = viewMode

        this.setViewMode(viewMode)
    }
}
</script>
