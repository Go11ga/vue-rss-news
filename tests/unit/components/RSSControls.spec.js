import { shallowMount, createLocalVue } from '@vue/test-utils'
import RSSControls from '@/components/RSSControls'
import Vuex from 'vuex'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('RSSControls.vue', () => {
    let wrapper
    let getters
    let mutations
    let store
    let toggleViewSpy
    let setItem
    let getItem

    beforeEach(() => {
        getters = {
            getActiveChannel: () => 'mos'
        }

        mutations = {
            setViewMode: jest.fn()
        }

        store = new Vuex.Store({
            getters,
            mutations
        })

        toggleViewSpy = jest.spyOn(RSSControls.methods, 'toggleView')
        setItem = jest.spyOn(Storage.prototype, 'setItem')
        getItem = jest.spyOn(Storage.prototype, 'getItem')

        wrapper = shallowMount(RSSControls, {
            localVue,
            store,
            stubs: ['router-link']
        })
    })

    afterEach(() => {
        wrapper.destroy()
    })

    it('правильно рендерится', () => {
        expect(wrapper.vm.$options.name).toMatch('RSSControls')

        expect(wrapper.findAll('.rss-controls__link').at(0).text()).toMatch('All')
        expect(wrapper.findAll('.rss-controls__link').at(1).text()).toMatch('Mos')
        expect(wrapper.findAll('.rss-controls__link').at(2).text()).toMatch('RIA')
    })

    it('snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
    })

    it('getter getActiveChannel', () => {
        expect(wrapper.find('.rss-controls__link--active').text()).toMatch('Mos')
    })

    it('клик на кнопке переключения отображения вида list', async () => {
        let btnList = wrapper.findAll('.rss-controls__btn').at(0)

        await btnList.trigger('click')

        expect(toggleViewSpy).toHaveBeenCalled()
        expect(toggleViewSpy).toHaveBeenCalledTimes(1)

        expect(wrapper.findAll('button').at(0).classes()[1]).toBe('rss-controls__btn--active')

        expect(setItem).toHaveBeenCalled()
        expect(setItem).toHaveBeenCalledTimes(1)
    })

    it('метод toggleView вызывает Vuex mutation setViewMode с payload', () => {
        wrapper.vm.toggleView('list')

        expect(mutations.setViewMode).toHaveBeenCalled()
        expect(mutations.setViewMode).toHaveBeenCalledTimes(1)
        expect(mutations.setViewMode).toHaveBeenCalledWith({}, 'list')

    })

    it('хук created', () => {
        expect(typeof RSSControls.created).toBe('function')

        expect(getItem).toHaveBeenCalled()

        expect(mutations.setViewMode).toHaveBeenCalled()
    })
})