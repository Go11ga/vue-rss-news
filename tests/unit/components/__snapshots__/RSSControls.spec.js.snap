// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RSSControls.vue snapshot 1`] = `
<div
  class="rss-controls"
>
  <div
    class="rss-controls__item"
  >
    <router-link-stub
      class="rss-controls__link"
      to="/all/1"
    >
      
                All
            
    </router-link-stub>
     
    <router-link-stub
      class="rss-controls__link rss-controls__link--active"
      to="/mos/1"
    >
      
                Mos
            
    </router-link-stub>
     
    <router-link-stub
      class="rss-controls__link"
      to="/ria/1"
    >
      
                RIA
            
    </router-link-stub>
  </div>
   
  <div
    class="rss-controls__item"
  >
    <button
      class="rss-controls__btn"
    >
      <svg
        class="rss-controls__icon"
      >
        <use
          xlink:href="#icon_list"
        />
      </svg>
    </button>
     
    <button
      class="rss-controls__btn rss-controls__btn--active"
    >
      <svg
        class="rss-controls__icon"
      >
        <use
          xlink:href="#icon_grid"
        />
      </svg>
    </button>
  </div>
</div>
`;
